Fix Collector configuration issues (#3443)
open-telemetry/opentelemetry.io: content/en/docs/collector/configuration.md

## Receivers <img width="35" class="img-initial" src="/img/logos/32x32/Receivers.svg"> {#receivers}

Receivers collect telemetry from one or more sources. They can be pull or push
based, and may support one or more [data sources](/docs/concepts/signals/).

> Receivers are configured in the `receivers:` section. Many receivers come with

Reviewer: I'd like to suggest that we drop the colon from the section name. The colon is an artifact of the YAML syntax. The section name is just the YAML key identifier. WDYT?

```suggestion
Receivers are configured in the `receivers` section. Many receivers come with
```

If you agree, this change will need to be applied elsewhere in the page.

  zipkin:
```

> For detailed receiver configuration, see the
> [receiver README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/README.md).

## Processors <img width="35" class="img-initial" src="/img/logos/32x32/Processors.svg"> {#processors}

> Processors take the data collected by receivers and modify or transform it

Reviewer 2: Not seeing a note in here about the ordering of processor matters, it's possible i missed it though

Author: Added in line 258.

before sending it to the exporters. Data processing happens according to rules
or settings defined for each processor, which might include filtering, dropping,
renaming, or recalculating telemetry, among other operations.

Processors are optional, although some
[are recommended](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor#recommended-processors).

> You can configure processors using the `processors:` section of the Collector

Reviewer: ```suggestion
You can configure processors using the `processors` section of the Collector
```

> configuration file. Some processors may require security settings. Any setting

Reviewer 2: The line about requiring security settings could use a link to either an example or be removed. In theory any components may require security settings in order to function :)

Author: Always in for _less_ docs when it makes sense. :) Removing.

      separator: '::'
```

> For detailed processor configuration, see the
> [processor README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/README.md).

## Exporters <img width="35" class="img-initial" src="/img/logos/32x32/Exporters.svg"> {#exporters}

Exporters send data to one or more backends or destinations. Exporters can be
pull or push based, and may support one or more
[data sources](/docs/concepts/signals/).

> The `exporters:` section contains exporters configuration. Most exporters

Reviewer: ```suggestion
The `exporters` section contains exporters configuration. Most exporters
```

  # Data sources: traces
  otlp/jaeger:

>     endpoint: jaeger.example.com:4317

Author: Yeah, I was unsure about which pattern to follow. I'm partial to using localhost everywhere if that's the most common use case in agent deployments. WDYT?

>     endpoint: jaeger.example.com:4317

Reviewer 3: Yeah, it might be confusing, so I would be OK with service.example.com:port

@open-telemetry/collector-approvers WDYT?

  # Data sources: metrics
  prometheusremotewrite:
    endpoint: http://prometheus.example.com:9411/api/prom/push
    # When using the official Prometheus (running via Docker)
    # endpoint: 'http://prometheus:9090/api/v1/write', add:
    # tls:
    #   insecure: true

  # Data sources: traces
  zipkin:
    endpoint: http://zipkin.example.com:9411/api/v2/spans
```

Notice that some exporters require x.509 certificates in order to establish
secure connections, as described in
[setting up certificates](#setting-up-certificates).

> For more information on exporter configuration, see the
> [exporter README.md](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/README.md).

## Connectors <img width="32" class="img-initial" src="/img/logos/32x32/Load_Balancer.svg"> {#connectors}

Connectors join two pipelines, acting as both exporter and receiver. A connector
consumes data as an exporter at the end of one pipeline and emits data as a
receiver at the beginning of another pipeline. The data consumed and emitted may
be of the same type or of different data types. You can use connectors to
summarize consumed data, replicate it, or route it.

> You can configure one or more connectors using the `connectors:` section of the

Reviewer: ```suggestion
You can configure one or more connectors using the `connectors` section of the
```

Collector configuration file. By default, no connectors are configured. Each
type of connector is designed to work with one or more pairs of data types and
may only be used to connect pipelines accordingly.

> Configuring a connector doesn't enable it. Connectors are enabled through
> pipelines within the [service](#service) section.

The following example shows the `count` connector and how it's configured in the

> `pipelines:` section. Notice that the connector acts as an exporter for traces

Reviewer: ```suggestion
`pipelines` section. Notice that the connector acts as an exporter for traces
```

          - 'attributes["env"] == "prod"'
          - 'name == "prodevent"'

service:
  pipelines:
    traces:
      receivers: [foo]
      exporters: [count]
    metrics:
      receivers: [count]
      exporters: [bar]
```

> For detailed connector configuration, see the
> [connector README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/connector/README.md).

## Extensions

Extensions are optional components that expand the capabilities of the Collector
to accomplish tasks not directly involved with processing telemetry data. For
example, you can add extensions for Collector health monitoring, service
discovery, or data forwarding, among others.

> You can configure extensions through the `extensions:` section of the Collector

Reviewer: ```suggestion
You can configure extensions through the `extensions` section of the Collector
```

    size_mib: 512
```

> For detailed extension configuration, see the
> [extension README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/extension/README.md).

## Service section {#service}

The service section is used to configure what components are enabled in the
Collector based on the configuration found in the receivers, processors,
exporters, and extensions sections. If a component is configured, but not
defined within the service section, then it's not enabled.

The service section consists of three subsections:

- `extensions:`
- `pipelines:`

> - `telemetry:`

Reviewer: Could we drop the colons here? I think that it is sufficient to name the YAML key. WDYT?

Author: Done! Switched to uppercase items so as to turn the list into merely structural, rather than YAML related.

### Extensions {#service-extensions}

> The `extensions:` subsection consists of a list of all extensions to enable. For

Reviewer: ```suggestion
The `extensions` subsection consists of a list of all extensions to enable. For
```

### Pipelines

> Pipelines are defined in the `pipelines:` subsection and can be of the following

Reviewer: ```suggestion
Pipelines are defined in the `pipelines` subsection and can be of the following
```

types:

> - `traces` collects and processes trace data.
> - `metrics` collects and processes metric data.
> - `logs` collects and processes log data.

Reviewer: ```suggestion
- `traces` collect and processes trace data.
- `metrics` collect and processes metric data.
- `logs` collect and processes log data.
```

> - `traces:` collects and processes trace data.
> - `metrics:` collects and processes metric data.
> - `logs:` collects and processes log data.

Reviewer: ```suggestion
- `traces` collects and processes trace data.
- `metrics` collects and processes metric data.
- `logs` collects and processes log data.
```

A pipeline consists of a set of receivers, processors and exporters. Before
including a receiver, processor, or exporter in a pipeline, make sure to define
its configuration in the appropriate section.

You can use the same receiver, processor, or exporter in more than one pipeline.
When a processor is referenced in multiple pipelines, each pipeline gets a
separate instance of the processor.

The following is an example of pipeline configuration. Note that the order of
processors dictates the order in which data is processed:

```yaml
service:
  pipelines:
    metrics:
      receivers: [opencensus, prometheus]
      processors: [batch]
      exporters: [opencensus, prometheus]
    traces:
      receivers: [opencensus, jaeger]
      processors: [batch, memory_limiter]
      exporters: [opencensus, zipkin]
```

### Telemetry

> You can configure telemetry for the Collector itself in the `telemetry:`
> subsection inside the `service:` section. Collector telemetry can be useful when

Reviewer: ```suggestion
You can configure telemetry for the Collector itself in the `telemetry`
subsection inside the `service` section. Collector telemetry can be useful when
```

troubleshooting Collector issues.

> The `logs:` subsection lets you configure logs generated by the Collector. By
> default, the Collector writes its logs to stderr with a log level of `INFO`. You
> can also add static key-value pairs to all logs using the `initial_fields`
> section. See the full list of `logs:` options

Reviewer: ```suggestion
The `logs` subsection lets you configure logs generated by the Collector. By
default, the Collector writes its logs to stderr with a log level of `INFO`. You
can also add static key-value pairs to all logs using the `initial_fields`
section. See the full list of `logs` options
```

[here.](https://github.com/open-telemetry/opentelemetry-collector/blob/7666eb04c30e5cfd750db9969fe507562598f0ae/config/service.go#L41-L97)

> The `metrics:` subsection lets you configure metrics generated by the Collector.

Reviewer: ```suggestion
The `metrics` subsection lets you configure metrics generated by the Collector.
```

By default, the Collector generates basic metrics about itself and expose them

> for scraping at `localhost:8888/metrics`. See the full list of options for the

Reviewer: With the following change the link checker will be ok:

```suggestion
for scraping at <http://localhost:8888/metrics>. See the full list of options for the
```

Author: Thanks!

> for scraping at `localhost:8888/metrics`. See the full list of options for the

Reviewer: Would it make sense for this to be a (clickable) link rather than use code format?

Author: I think the link checker doesn't like it. Any way of adding `localhost` as an exception?

Reviewer: See my followup suggestion. Marking this as resolved.

> `metrics:` subsection

Reviewer: ```suggestion
`metrics` subsection
```

> for scraping at <http://localhost:8888/metrics>. See the full list of options
> for the `metrics` subsection
> [here.](https://github.com/open-telemetry/opentelemetry-collector/blob/7666eb04c30e5cfd750db9969fe507562598f0ae/config/service.go#L99-L111)

Reviewer: As I mentioned in the other PR making changes to this section, I don't think we should be linking out to the source file. Especially because that [**config file has been deprecated**](https://github.com/open-telemetry/opentelemetry-collector/blob/50f013d0d242bf7b600eddf5cfe484e087c1de0c/CHANGELOG.md?plain=1#L1237)!

We should just list the "full options" (there aren't that many) here.

---

If you choose to keep the link for now, I'd at least suggest the following rewording to avoid a "here" link:

```suggestion
for scraping at <http://localhost:8888/metrics>. For the full list of options
for the `metrics` subsection, see the 
[`ServiceTelemetryMetrics struct`](https://github.com/open-telemetry/opentelemetry-collector/blob/7666eb04c30e5cfd750db9969fe507562598f0ae/config/service.go#L99-L111) comments.
```

Reviewer: @theletterf et al.: I've created a followup issue to track further improvements to this point:

- #3514

    namespace: $$DataVisualization
```

### Proxy support

Exporters that use the [`net/http`](https://pkg.go.dev/net/http) package respect
the following proxy environment variables:

- `HTTP_PROXY`: Address of the HTTP proxy
- `HTTPS_PROXY`: Address of the HTTPS proxy
- `NO_PROXY`: Addresses that must not use the proxy

If set at Collector start time, exporters, regardless of the protocol, proxy
traffic or bypass proxy traffic as defined by these environment variables.

### Authentication

Most receivers exposing an HTTP or gRPC port can be protected using the
Collector's authentication mechanism. Similarly, most exporters using HTTP or
gRPC clients can add authentication to outgoing requests.

The authentication mechanism in the Collector uses the extensions mechanism,
allowing for custom authenticators to be plugged into Collector distributions.
Each authentication extension has two possible usages:

- As client authenticator for exporters, adding auth data to outgoing requests
- As server authenticator for receivers, authenticating incoming connections.

For a list of known authenticators, see the
[Registry](/ecosystem/registry/?s=authenticator&component=extension). If you're

> interested in developing a custom authenticator, check out the
> ["Building a custom authenticator"](../custom-auth) document.

Reviewer: ```suggestion
interested in developing a custom authenticator, see 
[Building a custom authenticator](../custom-auth).
```

> To add a server authenticator to a receiver in the Collector, follow these steps
> :

Reviewer: ```suggestion
To add a server authenticator to a receiver in the Collector, follow these steps:
```

Strange that Prettier didn't complain about the lone colon on its own line.

        - otlp/auth
```

### Configuring certificates {#setting-up-certificates}

In a production environment, use TLS certificates for secure communication or
mTLS for mutual authentication. Follow these steps to generate self-signed
certificates as in this example. You might want to use your current cert
provisioning procedures to procure a certificate for production usage.

> Install [cfssl](https://github.com/cloudflare/cfssl) and create the following

Reviewer: Either use CFSSL or put in code font, e.g.:

```suggestion
Install [`cfssl`](https://github.com/cloudflare/cfssl) and create the following
```
